src/firewall/nft.rs

```diff
@@ -0,0 +1,1114 @@
+                            expr::Meta {
+                                key: expr::MetaKey::Mark,
+                            },
+                        ))),
+                        Box::new(expr::Expression::Number(MASK)),
+                    )),
+                })],
+            ));
+        }
+
+        // We need rules in Prerouting and Output pointing to our dnat chain.
+        // But only if they do not exist.
+        let match_jump_dnat = get_rule_matcher_jump_to(DNATCHAIN.to_string());
+        // Prerouting: fib daddr type local jump <dnat_chain>
+        // Output: fib daddr type local jump <dnat_chain>
+        let mut rules_hash: HashSet<expr::FibFlag> = HashSet::new();
+        rules_hash.insert(expr::FibFlag::Daddr);
+        let base_conditions: Vec<stmt::Statement> = vec![
+            stmt::Statement::Match(stmt::Match {
+                left: expr::Expression::Named(expr::NamedExpression::Fib(expr::Fib {
+                    result: expr::FibResult::Type,
+                    flags: rules_hash,
+                })),
+                right: expr::Expression::String("local".to_string()),
+                op: stmt::Operator::EQ,
+            }),
+            get_jump_action(DNATCHAIN),
+        ];
+        if get_matching_rules_in_chain(&existing_rules, PREROUTINGCHAIN, &match_jump_dnat)
+            .is_empty()
+        {
+            batch.add(make_rule(PREROUTINGCHAIN, base_conditions.clone()));
+        }
+        if get_matching_rules_in_chain(&existing_rules, OUTPUTCHAIN, &match_jump_dnat).is_empty() {
+            batch.add(make_rule(OUTPUTCHAIN, base_conditions.clone()));
+        }
+
+        // Forward chain: ct state invalid drop
+        let match_deny = |r: &schema::Rule| -> bool {
+            for statement in &r.expr {
+                match statement {
+                    stmt::Statement::Drop(_) => return true,
+                    _ => continue,
+                }
+            }
+            false
+        };
+        if get_matching_rules_in_chain(&existing_rules, FORWARDCHAIN, match_deny).is_empty() {
+            batch.add(make_rule(
+                FORWARDCHAIN,
+                vec![
+                    stmt::Statement::Match(stmt::Match {
+                        left: expr::Expression::Named(expr::NamedExpression::CT(expr::CT {
+                            key: "state".to_string(),
+                            family: None,
+                            dir: None,
+                        })),
+                        right: expr::Expression::String("invalid".to_string()),
+                        op: stmt::Operator::IN,
+                    }),
+                    stmt::Statement::Drop(None),
+                ],
+            ));
+        }
+
+        // Basic forwarding for all subnets
+        if let Some(nets) = network_setup.subnets {
+            for subnet in nets {
+                let chain = get_subnet_chain_name(subnet, false);
```

Comment:
I wonder given nft support both ipv4 and v6 at once if doing this by address is the wrong approach? Maybe we should switch to using the bridge interface as filter (iifname oifname)?

Then we only need ip addresses for the port forwarding rules.

Reply:
It feels like this might break IPv4/IPv6 selective enablement on the network (in that we'll have v6 on networks that otherwise might not, because they auto-assigned link-local addresses) but I'm not sure about that

Reply:
ah good point, not sure about link local but I think you are right there could be a future where we want to NAT ipv4 but not NAT the ipv6 space at the same time so I guess doing this by subnet is better as it gives us some flexibility in that regards